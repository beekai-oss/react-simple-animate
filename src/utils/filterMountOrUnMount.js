// @flow
type Result = {
  mappedChildren: Array<any>,
  willUnMount: boolean,
  willMount: boolean,
};

export default function filterMountOrUnMount(
  childrenStoreInState: Object | Array<React$Element<any>>,
  children: Object | Array<React$Element<any>>,
): Result {
  const result: Result = {
    mappedChildren: [],
    willUnMount: false,
    willMount: false,
  };
  const childrenShallowCopy = Array.isArray(children) ? children : [children];
  const childrenStoreInStateShallowCopy = Array.isArray(childrenStoreInState)
    ? childrenStoreInState
    : [childrenStoreInState];

  // any children got removed
  childrenStoreInStateShallowCopy.forEach((element: React$Element<any>) => {
    if (!Object.prototype.hasOwnProperty.call(element, 'key')) return;

    if (childrenShallowCopy.find(child => child.key === element.key)) {
      result.mappedChildren.push({
        ...element,
        willUnmount: false,
        willMount: false,
      });
    } else {
      result.willUnMount = true;
      result.mappedChildren.push({
        ...element,
        willUnmount: true,
      });
    }
  });

  // append missing child
  childrenShallowCopy.forEach((child, index) => {
    if (
      Object.prototype.hasOwnProperty.call(child, 'key') &&
      !result.mappedChildren.find(r => r.key === child.key)
    ) {
      result.willMount = true;
      result.mappedChildren.splice(index === 0 ? 0 : index + 1, 0, {
        ...child,
        willMount: true,
      });
    }
  });

  return result;
}
