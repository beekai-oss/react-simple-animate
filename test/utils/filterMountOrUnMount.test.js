import filterMountOrUnMount from '../../src/utils/filterMountOrUnmount';

describe('filterMountOrUnMount', () => {
  it('should mark child is about to be unMount', () => {
    const elements = [
      {
        key: 1,
      },
      {
        key: 2,
      },
    ];
    const children = [
      {
        key: 1,
      },
    ];

    expect(filterMountOrUnMount(elements, children)).toMatchSnapshot();
  });

  it('should append new children', () => {
    const elements = [
      {
        key: 1,
      },
      {
        key: 2,
      },
      {
        key: 3,
      },
    ];
    const children = [
      {
        key: 0,
      },
      {
        key: 1,
      },
      {
        key: 4,
      },
      {
        key: 6,
      },
      {
        key: 3,
      },
      {
        key: 5,
      },
    ];

    expect(filterMountOrUnMount(elements, children)).toMatchSnapshot();
  });
});
