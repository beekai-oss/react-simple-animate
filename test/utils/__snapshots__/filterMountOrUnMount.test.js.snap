// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterMountOrUnMount should append new children 1`] = `
Object {
  "childrenWillMount": true,
  "childrenWillUnmount": true,
  "mappedChildren": Array [
    Object {
      "key": 0,
      "willMount": true,
    },
    Object {
      "key": 1,
      "willMount": false,
      "willUnmount": false,
    },
    Object {
      "key": 2,
      "willUnmount": true,
    },
    Object {
      "key": 4,
      "willMount": true,
    },
    Object {
      "key": 6,
      "willMount": true,
    },
    Object {
      "key": 3,
      "willMount": false,
      "willUnmount": false,
    },
    Object {
      "key": 5,
      "willMount": true,
    },
  ],
}
`;

exports[`filterMountOrUnMount should mark child is about to be unMount 1`] = `
Object {
  "childrenWillMount": false,
  "childrenWillUnmount": true,
  "mappedChildren": Array [
    Object {
      "key": 1,
      "willMount": false,
      "willUnmount": false,
    },
    Object {
      "key": 2,
      "willUnmount": true,
    },
  ],
}
`;
